cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageTable.java

```diff
@@ -0,0 +1,484 @@
+import co.cask.cdap.api.lineage.field.ReadOperation;
+import co.cask.cdap.api.lineage.field.WriteOperation;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.proto.codec.OperationTypeAdapter;
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.cdap.proto.metadata.lineage.ProgramRunOperations;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.StructuredTableContext;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Fields;
+import co.cask.cdap.spi.data.table.field.Range;
+import co.cask.cdap.store.StoreDefinition;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset to store/retrieve field level lineage information.
+ */
+public class FieldLineageTable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(FieldLineageTable.class);
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Operation.class, new OperationTypeAdapter())
+    .create();
+
+  private static final String INCOMING_DIRECTION_MARKER = "i";
+  private static final String OUTGOING_DIRECTION_MARKER = "o";
+
+  private static final Type SET_FIELD_TYPE = new TypeToken<HashSet<String>>() { }.getType();
+  private static final Type SET_ENDPOINT_FIELD_TYPE = new TypeToken<HashSet<EndPointField>>() { }.getType();
+  private static final Type SET_OPERATION_TYPE = new TypeToken<HashSet<Operation>>() { }.getType();
+
+  private final StructuredTableContext structuredTableContext;
+  private StructuredTable endpointChecksumTable;
+  private StructuredTable operationsTable;
+  private StructuredTable destinationFieldsTable;
+  private StructuredTable summaryFieldsTable;
+
+  private FieldLineageTable(StructuredTableContext structuredTableContext) {
+    this.structuredTableContext = structuredTableContext;
+  }
+
+  /**
+   * Gets an instance of {@link FieldLineageTable}.
+   *
+   * @param structuredTableContext the {@link StructuredTableContext} for getting the table.
+   * @return an instance of {@link FieldLineageTable}
+   */
+  @VisibleForTesting
+  public static FieldLineageTable create(StructuredTableContext structuredTableContext) {
+    return new FieldLineageTable(structuredTableContext);
+  }
+
+  private StructuredTable getEndpointChecksumTable() {
+    if (endpointChecksumTable == null) {
```

Comment:
any reason for not creating these tables in the ctor itself?

Reply:
We want to get the table only when absolutely necessary - to avoid excess schema registry calls